---
title: "Class18"
author: "Philip Dai Le"
date: "11/26/2019", "editied 12/3/19"
output: html_document
---
##Cancer Genomics 

**Hands-on part 1, Section 1**
* Q1, How many Cases (i.e. patient samples) have been found to have p53 mutations?
  +4,008 cases affected by 1,291 mutations; 4251 cases affected by 638 CNV events
  
  
*Q2, What are the top 6 misssense mutations found in this gene?
HINT: Scroll down to the ‘TP53 - Protein’ section and mouse over the displayed plot. For example R175H is found in 156 cases.
  +R175H, R248Q, R273C, R273H, R282W, R248W
  
  
*Q3, Which domain of the protein (as annotated by PFAM) do these mutations reside in?
  +The mutations concentrate in the Pf00870, DNA-binding domain, of the protein
  
  
*Q4, What are the top 6 primary sites (i.e. cancer locations such as Lung, Brain, etc.) with p53 mutations and how many primary sites have p53 mutations been found in?
HINT: Clicking on the number links in the Cancer Distribution section will take you to a summary of available data accross cases, genes, and mutations for p53. Looking at the cases data will give you a ranked listing of primary sites.
  +Bronchus and lung, ovary, brain, breast, colon, and stomach are the top 6 sites
  
  
*Q5, What is the most frequentely mutated position associated with cancer in the KRas protein (i.e. the amino acid with the most mutations)?
  +The most mutated region would be PF00071, the GTPase protein
  
  
*Q6, Are KRas mutations common in Pancreatic Adenocarcinoma (i.e. is the Pancreas a common ‘primary site’ for KRas mutations?).
  +Out of the 839 cancer cases regarding KRas mutations, 182 cases of pancreatic cases have a KRas mutation in 137 of them.



*Q7, What is the ‘TGCA project’ with the most KRas mutations?
  +TCGA-PAAD, pancreas has the most KRas
  
  
*Q8,


** Hands-on, Section 2**

Loading in GenomicDataCommons, TCGAbiolinks, and maftools after installing via "BiocManager::install( c("GenomicDataCommons", "TCGAbiolinks", "maftools") )"
```{r}
library (GenomicDataCommons)
library (TCGAbiolinks)
library(maftools)
```

Checking the GDC status
```{r}
status()
```

Four main sets of metadata: projects(), cases(), files(), and annotations(). In this case, we are looking at the projects.
```{r}
projects <- getGDCprojects()
head(projects)
View(projects)
```

Looking at cases now.
```{r}
cases_by_project <-cases()  %>%
  
facet("project.project_id") %>%
  
aggregations()

head(cases_by_project)
View(cases_by_project)
```

Visualizing the case data via a barplot
```{r}
x<- cases_by_project$project.project_id
View(x)

colvec<-rep("lightblue", nrow(x)) #rep = replicate, will replicate the color light blue for each project
colvec [(x$key=="TCGA-PAAD")] <-"red"

barplot (x$doc_count, names.arg= x$key, log = "y", col=colvec, las=2)
```


Determining the number of cases and types of data associated with the given project
```{r}
library (TCGAbiolinks)
#samp <- getSampleFilesSummary ("TGGA-PAAD")
#head(samp)
```

```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
View(samp)
```

Focusing on a data type of interest from the cases by utilizing GDCquery(). Focus on "find all gene expression data files for all pancreatic cancer patients".
```{r}
query <- GDCquery(project="TCGA-PAAD",
                  data.category="Transcriptome Profiling",
                  data.type="Gene Expression Quantification")

ans <- getResults(query)
```

##Lecture 18, Hands-on part 2
Designing a specific cancer vaccine
```{r}
seqs<-read.fasta("lecture18_sequences.fa")
wt <- seqs$P53_wt
mut <-seqs$P53_mutant
seqs
```

```{r}
library (bio3d)
library (muscle)
```

Reading in aligned sequences of p53_wt and p53_mut
```{r}
#seq <- read.csv("muscleAlignment.clw") 
#align2 <-seqaln (seqs)

seqs <- seqaln(seqs)
#Tried to use seqaln, but need to install muscle via biocmanager. Had to move muscle.exe into class18 folder to use

```

Determining alignment differences 
```{r}
ide <- conserv(seqs$ali, method="identity")
#function outputs 1.0 for identical and 0.5 for mismatch
mutant.sites <- which(ide < 1)
mutant.sites #assign all mismatch 0.5's to mutant_sites

gaps <-gap.inspect(seqs)
mutant.sites <-mutant.sites [mutant.sites %in% gaps$f.inds]
#%in% filters out mutant sites
```

```{r}
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

